### PR TITLE
Nova 4

### PR DESCRIPTION
**Pull request by advoor**

### CODE REVIEW

File: src/NovaEditorJsConverter.php [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        $this->renderCallbacks[$block] = $callback;
+    }
+
+    /**
+     * Renders the given EditorJS data to safe HTML.
+     *
+     * @param iterable|object|string $data Data to render
+     * @return HtmlString           Safe, directly returnable string.
+     */
+    public function generateHtmlOutput(iterable|object|string $data): HtmlString
```
Comment (roelofr) [inline src/NovaEditorJsConverter.php]: Iterable is an object type, which makes it redundant

```suggestion
    public function generateHtmlOutput(object|string $data): HtmlString
```

File: src/NovaEditorJsData.php [hunk 2]
```diff
@@ -9,13 +9,31 @@
...
+    {
+        if (is_string($attributes)) {
+            $attributes = json_decode($attributes);
+        }
+
+        foreach ($attributes as $key => $value) {
+            $this->attributes[$key] = $value;
+        }
+    }
+
```
Comment (roelofr) [inline src/NovaEditorJsData.php:29]: This probably has to do with me writing [this line in NovaEditorJsCast][1] incorrectly(the third argument is depth, not flags).

[1]: https://github.com/advoor/nova-editor-js/pull/61/files#diff-7aa0ae91aecd77ae7f8cfb3647c1709df8f87373987dd37702f7a7969c823ceaR57

File: phpunit.xml [hunk 3]
```diff
@@ -1,24 +1,16 @@
...
-    </testsuites>
-    <filter>
-        <whitelist processUncoveredFilesFromWhitelist="true">
-            <directory suffix=".php">./src</directory>
-        </whitelist>
-    </filter>
-    <php>
-        <server name="APP_ENV" value="testing"/>
-    </php>
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" backupGlobals="false" backupStaticAttributes="false" bootstrap="vendor/autoload.php" colors="true" convertErrorsToExceptions="true" convertNoticesToExceptions="true" convertWarningsToExceptions="true" processIsolation="false" stopOnFailure="false" xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/9.3/phpunit.xsd">
```
Comment (roelofr) [inline phpunit.xml:2]: Too bad phpunit's migration wraps this, makes it hard to read 😕

File: src/NovaEditorJsConverter.php [hunk 4]
```diff
@@ -36,10 +36,10 @@ public function addRender(string $block, callable $callback): void
     /**
      * Renders the given EditorJS data to safe HTML.
      *
-     * @param iterable|object|string $data Data to render
-     * @return HtmlString           Safe, directly returnable string.
+     * @param $data
+     * @return \Illuminate\Support\HtmlString Safe, directly returnable string.
      */
-    public function generateHtmlOutput(iterable|object|string $data): HtmlString
+    public function generateHtmlOutput($data): HtmlString
```
Comment (roelofr) [inline src/NovaEditorJsConverter.php:42]: Could be `mixed`?

File: src/NovaEditorJsCast.php [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+     * @return NovaEditorJsData|null
+     */
+    public function get($model, string $key, $value, array $attributes)
+    {
+        if ($value === null) {
+            return null;
+        }
+
+        try {
+            return new NovaEditorJsData(json_decode($value, true, JSON_THROW_ON_ERROR));
```
Comment (roelofr) [inline src/NovaEditorJsCast.php:57]:
```suggestion
            return new NovaEditorJsData(json_decode($value, true, 512, JSON_THROW_ON_ERROR));
```

File: src/NovaEditorJsData.php (hunk 2 above)
Comment (advoor) [reply src/NovaEditorJsData.php:29]: Doesn't look like it, this still is receiving some type of string

File: src/NovaEditorJsConverter.php (hunk 4 above)
Comment (advoor) [reply src/NovaEditorJsConverter.php]: Good shout. Pushed

**Pull request merged**